File: scripts/agent-id-support-testing/1-ingest-two-agents.py
"""
Step 1: Ingest episodes from two different agents into the same project.

Agent "scout" discovers information about people and places.
Agent "analyst" analyzes relationships and draws conclusions.

Both agents mention "Alice" -- so the Alice entity should end up
with agent_ids=["scout", "analyst"] after dedup merges.
"""

import asyncio
import shutil
from datetime import datetime, timezone

# ruff: noqa: E402
import sys, os  # noqa: E401

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..'))

from shared import DB_PATH, make_graphiti

EPISODES = [
    # --- Scout agent's observations ---
    {
        'name': 'scout-obs-1',
        'body': 'Alice works at Acme Corp as a software engineer. She started last month.',
        'source': 'field report',
        'agent_id': 'scout',
    },
    {
        'name': 'scout-obs-2',
        'body': 'Bob is the CTO of Acme Corp. He founded the company in 2019.',
        'source': 'field report',
        'agent_id': 'scout',
    },
    {
        'name': 'scout-obs-3',
        'body': 'Acme Corp is headquartered in Denver, Colorado.',
        'source': 'field report',
        'agent_id': 'scout',
    },
    # --- Analyst agent's conclusions ---
    {
        'name': 'analyst-note-1',
        'body': 'Alice reports to Bob at Acme Corp. She is on the infrastructure team.',
        'source': 'analysis',
        'agent_id': 'analyst',
    },
    {
        'name': 'analyst-note-2',
        'body': 'Carol is a product manager at Acme Corp. She works closely with Alice.',
        'source': 'analysis',
        'agent_id': 'analyst',
    },
]


async def main():
    # Wipe any previous test DB (Kuzu may create a file or directory)
    if os.path.exists(DB_PATH):
        print(f'Removing old test database at {DB_PATH}')
        if os.path.isdir(DB_PATH):
            shutil.rmtree(DB_PATH)
        else:
            os.remove(DB_PATH)
    # Also remove the .lock file Kuzu sometimes creates
    lock_path = DB_PATH + '.lock'
    if os.path.exists(lock_path):
        os.remove(lock_path)

    g = await make_graphiti()
    now = datetime.now(timezone.utc)

    print(f'\nIngesting {len(EPISODES)} episodes (group_id defaults to Kuzu default)...\n')

    for i, ep in enumerate(EPISODES):
        print(f'  [{i + 1}/{len(EPISODES)}] agent={ep["agent_id"]!r:12s} name={ep["name"]!r}')
        result = await g.add_episode(
            name=ep['name'],
            episode_body=ep['body'],
            source_description=ep['source'],
            reference_time=now,
            agent_id=ep['agent_id'],
        )
        print(f'           -> {len(result.nodes)} nodes, {len(result.edges)} edges extracted')

    print('\nDone! Database saved to:', DB_PATH)
    print('Run 2-search-with-agent-filter.py next.')


if __name__ == '__main__':
    asyncio.run(main())

File: cpp_refactor/examples/agent_attribution/1_ingest_two_agents.cpp
/*
 * Step 1: Ingest episodes from two different agents into the same project.
 *
 * Agent "scout" discovers information about people and places.
 * Agent "analyst" analyzes relationships and draws conclusions.
 *
 * Both agents mention "Alice" -- so the Alice entity should end up
 * with agent_ids=["scout", "analyst"] after dedup merges.
 */

#include "shared.h"

#include <chrono>
#include <filesystem>
#include <format>

struct Episode {
    std::string name;
    std::string body;
    std::string source;
    std::string agent_id;
};

int main() {
    std::vector<Episode> episodes = {
        // --- Scout agent's observations ---
        {"scout-obs-1",
         "Alice works at Acme Corp as a software engineer. She started last month.",
         "field report", "scout"},
        {"scout-obs-2",
         "Bob is the CTO of Acme Corp. He founded the company in 2019.",
         "field report", "scout"},
        {"scout-obs-3",
         "Acme Corp is headquartered in Denver, Colorado.",
         "field report", "scout"},
        // --- Analyst agent's conclusions ---
        {"analyst-note-1",
         "Alice reports to Bob at Acme Corp. She is on the infrastructure team.",
         "analysis", "analyst"},
        {"analyst-note-2",
         "Carol is a product manager at Acme Corp. She works closely with Alice.",
         "analysis", "analyst"},
    };

    // Wipe any previous test DB
    auto db = shared::db_path();
    if (std::filesystem::exists(db)) {
        std::cout << "Removing old test database at " << db << "\n";
        std::filesystem::remove_all(db);
    }
    // Also remove .lock file Kuzu sometimes creates
    auto lock_path = db + ".lock";
    if (std::filesystem::exists(lock_path)) {
        std::filesystem::remove(lock_path);
    }

    auto g = shared::make_graphiti();
    auto now = std::chrono::system_clock::now();

    std::cout << std::format("\nIngesting {} episodes (group_id=\"\")...\n\n", episodes.size());

    for (size_t i = 0; i < episodes.size(); ++i) {
        auto& ep = episodes[i];
        std::cout << std::format("  [{}/{}] agent={:12s} name=\"{}\"\n",
            i + 1, episodes.size(), ep.agent_id, ep.name);

        auto result = g.add_episode({
            .name = ep.name,
            .body = ep.body,
            .source_description = ep.source,
            .reference_time = now + std::chrono::seconds(i * 60), // stagger times
            .agent_id = ep.agent_id,
        });

        if (result.has_value()) {
            auto& r = result.value();
            std::cout << std::format("           -> {} nodes, {} edges extracted\n",
                r.nodes.size(), r.edges.size());
        } else {
            std::cout << std::format("           -> ERROR: {}\n", result.error().message);
        }
    }

    std::cout << "\nDone! Database saved to: " << db << "\n";
    std::cout << "Run 2_search_with_agent_filter next.\n";
    return 0;
}
